File: backend/services/coding_languages.py
"""Shared coding language registry (domains, editor, runner, AI)."""

from __future__ import annotations

from typing import Any

# id -> metadata used across API, runner, and AI prompts
CODING_LANGUAGES: dict[str, dict[str, Any]] = {
    "python": {
        "label": "Python",
        "extension": "py",
        "entry": "main.py",
        "monaco": "python",
        "runnable": True,
    },
    "javascript": {
        "label": "JavaScript",
        "extension": "js",
        "entry": "main.js",
        "monaco": "javascript",
        "runnable": True,
    },
    "typescript": {
        "label": "TypeScript",
        "extension": "ts",
        "entry": "main.ts",
        "monaco": "typescript",
        "runnable": True,
    },
    "java": {
        "label": "Java",
        "extension": "java",
        "entry": "Solution.java",
        "monaco": "java",
        "runnable": True,
    },
    "cpp": {
        "label": "C++",
        "extension": "cpp",
        "entry": "main.cpp",
        "monaco": "cpp",
        "runnable": True,
    },
    "csharp": {
        "label": "C#",
        "extension": "cs",
        "entry": "Solution.cs",
        "monaco": "csharp",
        "runnable": False,
    },
    "go": {
        "label": "Go",
        "extension": "go",
        "entry": "main.go",
        "monaco": "go",
        "runnable": True,
    },
    "ruby": {
        "label": "Ruby",
        "extension": "rb",
        "entry": "main.rb",
        "monaco": "ruby",
        "runnable": True,
    },
    "php": {
        "label": "PHP",
        "extension": "php",
        "entry": "main.php",
        "monaco": "php",
        "runnable": True,
    },
    "kotlin": {
        "label": "Kotlin",
        "extension": "kt",
        "entry": "Solution.kt",
        "monaco": "kotlin",
        "runnable": False,
    },
    "rust": {
        "label": "Rust",
        "extension": "rs",
        "entry": "main.rs",
        "monaco": "rust",
        "runnable": False,
    },
    "swift": {
        "label": "Swift",
        "extension": "swift",
        "entry": "main.swift",
        "monaco": "swift",
        "runnable": False,
    },
}

SUPPORTED_LANGUAGES = frozenset(CODING_LANGUAGES.keys())


def language_label(lang: str) -> str:
    meta = CODING_LANGUAGES.get((lang or "").strip().lower())
    return meta["label"] if meta else lang


def language_entry(lang: str) -> str:
    meta = CODING_LANGUAGES.get((lang or "").strip().lower())
    return meta["entry"] if meta else "main.py"


def language_extension(lang: str) -> str:
    meta = CODING_LANGUAGES.get((lang or "").strip().lower())
    return meta["extension"] if meta else "txt"


def languages_for_api() -> list[dict[str, Any]]:
    return [
        {
            "id": key,
            "label": meta["label"],
            "extension": meta["extension"],
            "runnable": bool(meta.get("runnable")),
        }
        for key, meta in CODING_LANGUAGES.items()
    ]

File: frontend/src/lib/coding-languages.ts
import type { CodingLanguage } from '@/types/api'

/** Languages available for domains / Monaco / AI generate. */
export const CODING_LANGUAGE_OPTIONS: {
  id: CodingLanguage
  label: string
  entry: string
  monaco: string
}[] = [
  { id: 'python', label: 'Python', entry: 'main.py', monaco: 'python' },
  { id: 'java', label: 'Java', entry: 'Solution.java', monaco: 'java' },
  { id: 'javascript', label: 'JavaScript', entry: 'main.js', monaco: 'javascript' },
  { id: 'typescript', label: 'TypeScript', entry: 'main.ts', monaco: 'typescript' },
  { id: 'cpp', label: 'C++', entry: 'main.cpp', monaco: 'cpp' },
  { id: 'csharp', label: 'C#', entry: 'Solution.cs', monaco: 'csharp' },
  { id: 'go', label: 'Go', entry: 'main.go', monaco: 'go' },
  { id: 'ruby', label: 'Ruby', entry: 'main.rb', monaco: 'ruby' },
  { id: 'php', label: 'PHP', entry: 'main.php', monaco: 'php' },
  { id: 'kotlin', label: 'Kotlin', entry: 'Solution.kt', monaco: 'kotlin' },
  { id: 'rust', label: 'Rust', entry: 'main.rs', monaco: 'rust' },
  { id: 'swift', label: 'Swift', entry: 'main.swift', monaco: 'swift' },
]

export function codingLanguageMeta(id: string) {
  return (
    CODING_LANGUAGE_OPTIONS.find((o) => o.id === id) ?? CODING_LANGUAGE_OPTIONS[0]
  )
}

export function defaultEntryForLanguage(language: string) {
  return codingLanguageMeta(language).entry
}

export function monacoLanguageFor(path: string, language: string) {
  if (path.endsWith('.ts') || path.endsWith('.tsx')) return 'typescript'
  if (path.endsWith('.js') || path.endsWith('.jsx')) return 'javascript'
  if (path.endsWith('.json')) return 'json'
  if (path.endsWith('.md')) return 'markdown'
  if (path.endsWith('.java')) return 'java'
  if (path.endsWith('.cpp') || path.endsWith('.cc') || path.endsWith('.h')) return 'cpp'
  if (path.endsWith('.cs')) return 'csharp'
  if (path.endsWith('.go')) return 'go'
  if (path.endsWith('.rb')) return 'ruby'
  if (path.endsWith('.php')) return 'php'
  if (path.endsWith('.kt')) return 'kotlin'
  if (path.endsWith('.rs')) return 'rust'
  if (path.endsWith('.swift')) return 'swift'
  if (path.endsWith('.py')) return 'python'
  return codingLanguageMeta(language).monaco
}
